File: aln-wrench-qt/aln/frame.cpp
#include "ax25frame.h"

QByteArray toFrameBuffer(QByteArray content) {
    return toFrameBuffer(content.data(), content.size());
}

QByteArray toFrameBuffer(const char* content, int len) {
    QByteArray ary;
    QBuffer buffer(&ary);
    buffer.open(QIODevice::Append);
    for (int i = 0; i < len; i++) {
        char b = content[i];
        if (b == End){
            buffer.write(&Esc, 1);
            buffer.write(&EndT, 1);
        } else if (b == Esc) {
            buffer.write(&Esc, 1);
            buffer.write(&EscT, 1);
        } else {
            buffer.write(&b, 1);
        }
    }
    buffer.write(&End, 1);
    buffer.close();
    return ary;
}

File: java-android/app/src/main/java/biglittleidea/aln/Frame.java
package biglittleidea.aln;

import java.nio.ByteBuffer;

public class Frame {
    public static int BufferSize = 4096;
    public static final byte End = (byte)0xC0;
    public static final byte Esc = (byte)0xDB;
    public static final byte EndT = (byte)0xDC;
    public static final byte EscT = (byte)0xDD;

    // Creates a byte array containing the serialized packet framed for transmission
    public static byte[] toAX25Buffer(byte[] content) {
        ByteBuffer buffer = ByteBuffer.allocate(BufferSize);
        for (int i = 0; i < content.length; i++) {
            Byte b = content[i];
            if (b == End){
                buffer.put(Esc);
                buffer.put(EndT);
            } else if (b == Esc) {
                buffer.put(Esc);
                buffer.put(EscT);
            } else {
                buffer.put(b);
            }
        }
        buffer.put(End);
        byte[] data = new byte[buffer.position()];
        buffer.rewind();
        buffer.get(data);
        return data;

    }

}
